### PR TITLE
Collapse LinalgExt into the main source tree

### PR DESCRIPTION
**Pull request by hanhanW**
The revision moves LinalgExt to compiler/Dialect, and fixes styles. The python binding support is removed. It does not trigger any issues at this moment. We can add it back if needed.

It retires an "unused" LinalgExt op (i.e.,`do_not_dce_operands`), which is only used by TransformInterpreter. The TransformInterpreter.cpp should be deprecated soon because we already have transform dialect setup in IREE main tree.

The step to generate the revision:

https://github.com/openxla/iree/pull/16407/commits/502ca323af981ed8b2313fa958b3002b04115f6f: Retire LinalgExt do_not_dce_operands op.

https://github.com/openxla/iree/pull/16407/commits/fddc0e51546e280c603d68070d6fee9cc7149be3: Run `git mv` to move files; manually create BILD.bazel

https://github.com/openxla/iree/pull/16407/commits/66ce8a2d84888ad0962221e9bf5e02c17cf62c1e: Remove LinalgExt from `llvm-external-projects/iree-dialects`.

https://github.com/openxla/iree/pull/16407/commits/37be17a27ff3258ad4763828ce9bfd214d752136 Update BUILD.bazel files with below commands:

```
sed -i '' -e "s/llvm-external-projects\/iree-dialects:IREELinalgExtDialect/compiler\/src\/iree\/compiler\/Dialect\/LinalgExt\/IR/g" **/BUILD.bazel
sed -i '' -e "s/llvm-external-projects\/iree-dialects:IREELinalgExtTransforms/compiler\/src\/iree\/compiler\/Dialect\/LinalgExt\/Transforms/g" **/BUILD.bazel
sed -i '' -e "s/llvm-external-projects\/iree-dialects:IREELinalgExtUtils/compiler\/src\/iree\/compiler\/Dialect\/LinalgExt\/Utils/g" **/BUILD.bazel
sed -i '' -e "s/llvm-external-projects\/iree-dialects:IREELinalgExtPasses/compiler\/src\/iree\/compiler\/Dialect\/LinalgExt\/Transforms/g" **/BUILD.bazel
sed -i '' -e "s/llvm-external-projects\/iree-dialects:IREELinalgExtTransformOps/compiler\/src\/iree\/compiler\/Dialect\/LinalgExt\/TransformExtensions:LinalgExtExtensions/g" **/BUILD.bazel
```

https://github.com/openxla/iree/pull/16407/commits/a8b6dd0de47b28e88eddaca1b6a1c07d10b82f0d: Manually fix BUILD and CMakeLists.txt
- They are not modeled by bazel_to_cmake.py
- LinalgExt/Transforms/BUILD.bazel missed a dep

https://github.com/openxla/iree/pull/16407/commits/e89ac64c73524527b62df0a3ec1d7cbc73309699: Run below commands to fix includes.

```
sed -i '' -e "s/iree-dialects\/Dialect\/LinalgExt/iree\/compiler\/Dialect\/LinalgExt/g" **/*.td
sed -i '' -e "s/iree-dialects\/Dialect\/LinalgExt/iree\/compiler\/Dialect\/LinalgExt/g" **/*.h
sed -i '' -e "s/iree-dialects\/Dialect\/LinalgExt/iree\/compiler\/Dialect\/LinalgExt/g" **/*.cpp
sed -i '' -e "s/Dialect\/LinalgExt\/Passes/Dialect\/LinalgExt\/Transforms/g" **/*.cpp
sed -i '' -e "s/Dialect\/LinalgExt\/Passes/Dialect\/LinalgExt\/Transforms/g" **/*.h
sed -i '' -e "s/Dialect\/LinalgExt\/TransformOps\/LinalgExtTransformOps/Dialect\/LinalgExt\/TransformExtensions\/LinalgExtExtensionsOps/g" **/*.h
sed -i '' -e "s/Dialect\/LinalgExt\/TransformOps\/LinalgExtTransformOps/Dialect\/LinalgExt\/TransformExtensions\/LinalgExtExtensionsOps/g" **/*.cpp
```

https://github.com/openxla/iree/pull/16407/commits/c10e8ef5a7ee70fd155426915915c4bf3d9b1e37: Replace iree-dialects-opt with iree-opt for LinalgExt tests

https://github.com/openxla/iree/pull/16407/commits/37a315b3452f63239a29f122f99de307f937a910: Add iree. prefix to LinalgExtExtensionsOps.td and fix tests

https://github.com/openxla/iree/pull/16407/commits/f236ddb993ff1784ce5db4a653870b7f95d98307: Run `buildifier compiler/**/BUILD.bazel`

https://github.com/openxla/iree/pull/16407/commits/1d10a1ebd29c9199e4f88c5d6bd8510730d844c7: Run clang-format to fix headers

https://github.com/openxla/iree/pull/16407/commits/cdbf3ead407f49b9b45781314cadbb81d28a7002: Fixed indef in headers with below commands.

```
sed -i '' -e "s/IREE_DIALECTS/IREE_COMPILER/g" **/*.h
```

Manually fixes the style:

- https://github.com/openxla/iree/pull/16407/commits/4f8cf66948192f4d96e5110f8ec37fa65ef23551: Add braces on simple single-statement bodies of if/else/loop
- https://github.com/openxla/iree/pull/16407/commits/ee6236f4d28da26c9269e827665cc0ecc160de53: Changing compiler namespaces to the new single-line syntax.
- https://github.com/openxla/iree/pull/16407/commits/61e0fa9d222a307dfade1b654846a9dea03272af: Use structured binding for llvm::enumerate
- https://github.com/openxla/iree/pull/16407/commits/be5327d2b305b1249db75e9bab2e18cc0879d21d:  Replace llvm::zip with llvm::zip_equal

### CODE REVIEW

File: docs/website/postprocess_dialect_docs.py [hunk 1]
```diff
@@ -75,7 +75,7 @@ def main(args):
         "HALLoader.md": "compiler/src/iree/compiler/Modules/HAL/Loader/IR",
         "IOParameters.md": "compiler/src/iree/compiler/Modules/IO/Parameters/IR",
         "IREEInput.md": "llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/Input",
-        "IREELinalgExt.md": "llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR",
+        "IREELinalgExt.md": "compiler/src/iree/compiler/Dialect/LinalgExt/IR",
```
Comment (ScottTodd) [inline docs/website/postprocess_dialect_docs.py]: You can also move LinalgExt from "public dialects" to "internal dialects" here: https://github.com/openxla/iree/blob/da982154aebccb41c1cf9bf5594097a2e6906b19/docs/website/mkdocs.yml#L164-L168

That affects the categories on https://iree.dev/reference/ (maybe we should just fully collapse that now - projects like Turbine use more than the "public" dialects)

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/Transforms.h [hunk 2]
```diff
@@ -4,27 +4,18 @@
...
-class LinalgOp;
-}
-
-namespace iree_compiler {
-namespace IREE {
-namespace LinalgExt {
+namespace mlir::iree_compiler::IREE::LinalgExt {
 
 /// Pattern to rewrite a scf::ForallOp to the async dialect.
 struct ForallOpToAsyncRewriter : public OpRewritePattern<scf::ForallOp> {
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/Transforms.h:21]: Async dialect? Is IREE using that?

Do we know if some of this code is unused? Could push out to https://github.com/iree-org/iree-samples/tree/main/transform_dialect or elsewhere if so.

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp [hunk 3]
```diff
@@ -2485,41 +2509,53 @@ LogicalResult AttentionOp::verify() {
...
       return op->emitOpError("query and key head dimension mismatch");
+    }
   }
   if (numOperands == 6) {
     // Tiled/Flash attention.
     ShapedType maxType = *getMaxType();
     ShapedType sumType = *getSumType();
-    if (failed(checkShapeRank(op, "max", maxType, 1)))
+    if (failed(checkShapeRank(op, "max", maxType, 1))) {
       return failure();
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp:2538]: For some of these diffs, I'm wondering if an in-place cleanup PR (modifying the files while they are still in their old location) would be easier to review. As this PR is, the diffs are split across many commits.

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  NAME
+    LinalgExtOpsIncGen
+  TD_FILE
+    "LinalgExtOps.td"
+  OUTS
+    --gen-op-decls LinalgExtOps.h.inc
+    --gen-op-defs LinalgExtOps.cpp.inc
+    --dialect=iree_linalg_ext --gen-dialect-decls LinalgExtDialect.h.inc
+    --dialect=iree_linalg_ext --gen-dialect-defs LinalgExtDialect.cpp.inc
+)
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/LinalgExt/IR/CMakeLists.txt:98]: May want a `iree_tablegen_doc` in this file.

File: compiler/src/iree/compiler/Dialect/LinalgExt/BUILD.bazel
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/LinalgExt/BUILD.bazel:1]: lol @ the git file association here

image

may want to break that link to keep history cleaner

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtInterfaces.cpp [hunk 5]
```diff
@@ -26,16 +26,16 @@ IREE::LinalgExt::detail::verifyLinalgExtOpInterface(Operation *op) {
...
-#include "iree-dialects/Dialect/LinalgExt/IR/LinalgExtOpInterfaces.cpp.inc" // IWYU pragma: export
+#include "iree/compiler/Dialect/LinalgExt/IR/LinalgExtInterfaces.cpp.inc" // IWYU pragma: export
 
 template <typename Ty, typename DimOpTy>
 static void getDimValues(OpBuilder &b, Location loc, Value v, Ty t,
                          SmallVector<OpFoldResult> &dimVals) {
-  for (auto dim : llvm::enumerate(t.getShape())) {
-    if (ShapedType::isDynamic(dim.value())) {
-      dimVals.push_back(b.create<DimOpTy>(loc, v, dim.index()).getResult());
+  for (const auto &[idx, val] : llvm::enumerate(t.getShape())) {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtInterfaces.cpp]: Dont need `const &` . Its just a value.

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp [hunk 6]
```diff
@@ -361,8 +367,8 @@ LogicalResult ScatterOp::generateScalarImplementation(OpBuilder &b,
   auto updateIvs = ivs.drop_front(1);
 
   int64_t offset = starts.size() - updateIvs.size();
-  for (auto it : llvm::enumerate(updateIvs)) {
-    starts[it.index() + offset] = it.value();
+  for (const auto &[idx, iv] : llvm::enumerate(updateIvs)) {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp]: Nit: here and everywhere. Dont need `const &` for values.

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/BUILD.bazel [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        ":PassesIncGen",
+        "//compiler/src/iree/compiler/Dialect/LinalgExt/IR",
+        "//compiler/src/iree/compiler/Dialect/LinalgExt/Utils",
+        "//llvm-external-projects/iree-dialects:IREEInputDialect",
+        "@llvm-project//llvm:Support",
+        "@llvm-project//mlir:AffineDialect",
+        "@llvm-project//mlir:AffineUtils",
+        "@llvm-project//mlir:ArithDialect",
+        "@llvm-project//mlir:ArithUtils",
+        "@llvm-project//mlir:AsyncDialect",
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/BUILD.bazel:58]: I think `Async` dialect was pulled in for some experimentation. We should be able to just drop it.

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/Transforms.h (hunk 2 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/Transforms.h:21]: Should be able to delete this op and everything that relies on it :)

**Review by MaheshRavishankar — APPROVED** [review]
Looked through as carefully as I could. Mostly looks manual. The async stuff can be removed after this lands. Thanks so much for doing this!

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/Transforms.h (hunk 2 above)
Comment (ScottTodd) [reply compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/Transforms.h:21]: > The async stuff can be removed after this lands.

I'd really like to believe that we can sequence the work in that way, but I'm skeptical and don't want to adopt a bunch of tech debt in the core project if we can avoid it. How feasible would it be to do that sort of clean up prior to the code move?

Comment (hanhanW) [reply compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/Transforms.h:21]: I'm pretty sure that the async dialect can be removed soon. This op is actually not used by anyone. It is mostly for experiments in the past. I tried to remove them in https://github.com/openxla/iree/pull/12756 but people seemed to think they are useful.

IMO, this is the only tech debt about LinalgExt and we can delete it soon. IREE depends on LinalgExt, which depends on async dialect. It is not adding more deps on IREE; it just exposed the hidden dependency. Can I delete the exposed dep in a follow-up?

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp (hunk 3 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp:2538]: I was expecting reviewers to take a look at https://github.com/openxla/iree/pull/16407/commits/4f8cf66948192f4d96e5110f8ec37fa65ef23551

Dealing these with Github is very painful, so I put several commits in the PR.

File: compiler/src/iree/compiler/Dialect/LinalgExt/BUILD.bazel
Comment (hanhanW) [reply compiler/src/iree/compiler/Dialect/LinalgExt/BUILD.bazel:1]: Is it doable? I created the file in https://github.com/openxla/iree/commit/fddc0e51546e280c603d68070d6fee9cc7149be3. And there are no `iree_linalg_ext.py` file in the commit at all. I think this is resolved by Github or git, and it is probably not actionable...?

Comment (ScottTodd) [reply compiler/src/iree/compiler/Dialect/LinalgExt/BUILD.bazel:1]: Hmm.. I'm not finding a way to signal this to git. Seems like it just looks as the file diffs within each commit to determine what it thinks was a move vs a delete/new.

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/Transforms.h (hunk 2 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/Transforms.h:21]: I can also drop it in a commit, if you think it is better.

**Comment by hanhanW** [comment]
I'm going to merge main branch into this PR and land it tomorrow, if no more comments.

**Comment by ScottTodd** [comment]
> I'm going to merge main branch into this PR and land it tomorrow, if no more comments.

SGTM. I might need to poke the dialect doc generation for the website. Maybe I'll check out your branch and check it myself (no CI presubmit coverage and it's a bit tricky to test if you haven't built the website with generated files before).

Thanks for doing this!

**Comment by hanhanW** [comment]
> Maybe I'll check out your branch and check it myself (no CI presubmit coverage and it's a bit tricky to test if you haven't built the website with generated files before).

Thanks, Scott!

**Comment by ScottTodd** [comment]
> > Maybe I'll check out your branch and check it myself (no CI presubmit coverage and it's a bit tricky to test if you haven't built the website with generated files before).
> 
> Thanks, Scott!

Sent you https://github.com/hanhanW/iree/pull/2

**Review by ScottTodd — APPROVED** [review]

**Comment by MaheshRavishankar** [comment]
Quite a lift! Thanks @hanhanW

**Pull request merged**